### PR TITLE
feat(material/chips): prevent chips from being deleted when user holds backspace

### PR DESCRIPTION
**Pull request by gioragutt**
* By default, while holding backspace when the input is focused and empty, the last chip (if exists) would be focused.
* This makes is so that when people delete the content they wrote, they will eventually start deleting chips.
* This PR makes it so that only after the user releases the backspace after deleting all the content, subsequent backspace presses will focus the chips.

Edge Cases:
- [x] BS on initial empty input should focus last chip.
- [x] BS on initial non-empty input should not focus last chip (and not created new one due to focus out).
- [x] BS after changing input (from empty state or otherwise) should not focus last input or create chip.
- [x] BS after creating new chip should focus last input.

Fixes: #18659

### CODE REVIEW

**Comment by gioragutt** [comment]
@mmalerba I've come across a pretty annoying edge case.

It might be due to a shitty design (the flag), but maybe not.

Currently, from within the `chip-list` itself, I don't see a way to actually know when a chip was created (via one of the separator keys, focusout, or any other way).

This makes it so that when a chip is created, I don't flip the flag back to true, and a double backspace press is needed.

I'll illustrate:

* Focus input
* Write something
* Press `TAB`. This triggers focusout, creates a new chip, and focuses the next element in the page.
* Press `SHIFT` + `TAB`. This focuses back on the input
* Press `BACKSPACE`. Nothing happens (this is where the flag is flipped)
* Press `BACKSPACE`. Last chip is focused.

I would love for a review of the current method I'm using (the boolean flag).
If you think there is any other way, which is less quirky, I'd love to know (even if you can't think of one, just tell me there's a better way).

I can expose an observable from `MatChipTextControl` if needed, but at the same time I don't know if this is a smart idea to add it there (idk what else might [be able to] use it).

Currently I don't see this class exposed via public api in the normal chips, but I see it's exposed in the MDC code.

**Comment by gioragutt** [comment]
@mmalerba perhaps this entire behaviour should be moved to and managed by `chip-input`?

**Comment by mmalerba** [comment]
I think the flag approach looks fine, but yeah I would suggest moving all of this logic into the input. If that turns out to be too difficult for some reason, it should be ok to add a new observable to `MatChipTextControl`

**Comment by gioragutt** [comment]
@mmalerba good to hear. I’ll work on that today 👌🏻 Neither of the methods should be hard, I’d rather do the cleanest approach

**Comment by gioragutt** [comment]
@mmalerba I have a question, don't quite know how to tackle this.

When the user creates a new chip (via one of the separator keys, blur, etc), it's up to them to reset the input value,

As shown [in the chips demo in dev-app](https://github.com/angular/components/blob/master/src/dev-app/chips/chips-demo.ts#L73):

```ts
  add(event: MatChipInputEvent): void {
    const {input, value} = event;

    // Add our person
    if ((value || '').trim()) {
      this.people.push({ name: value.trim() });
    }

    // Reset the input value
    if (input) {
      input.value = '';
    }
  }
```

I've tried getting notified about when this happens (the `input.value = ''`) but I can't quite make it.
It does not raise an `input` event, so it does not trigger `_onInput()`, and I also tried to subscribe to `fromEvent(this._inputElement, 'input')` and got nothing.

This is a problem we have regardless of whether the code is in `chip-input` or `chip-list`, and also if the user just decides to do the same on his own (as in - get reference to the input and mess with the input value).

Do you have any thoughts on how we can do that?

**Comment by mmalerba** [comment]
Hmmm yeah this is a tricky one. What if in the chip-list you listen to changes in the child chips and call a method on the chip input to reset the flag whenever the children change?

**Comment by gioragutt** [comment]
@mmalerba problem here is that it's not about whether the chip was actually added to the list or not, but whether the input value was changed by messing with the dom element itself.

I'm thinking about all kinds of RxJS flows I can do with the input events and whatnot, but from my testing - this change does not raise an event, as I stated above.

Any chance there's another component in the library that had to deal with a similar issue? Is there someone I can ask, or perhaps you can give me directions to what kind of behaviour I should look for in other components to find that myself?

**Comment by mmalerba** [comment]
The only way I can think of to detect when the input's value is programatically set would be to use `Object.defineProperty` to define a setter for the `value` property. However, this is not a pattern we currently use in our library and I don't think we want to start doing that.

You're right that resetting it when the chips change isn't exactly right, but in practice when the user types a separator character the input will be reset and a chip will be added, so it does cover the most common case.

To cover other edge cases people might have I would just add a `clear` method that sets the value to `''` and resets the flag. You can update our examples to use that method so we guide people down the right path.

Does this sound like a good approach to you?

**Comment by gioragutt** [comment]
Hey @mmalerba, I know it's been a while, hope you're doing fine (Covid and all).

The issue (#18659) got some comments these last few days so I decided to see what was left in the PR and fix it.

File: src/material/chips/chip-input.ts
Comment (mmalerba) [inline src/material/chips/chip-input.ts:35]: rather than having the `clearInput` method on the event, how about adding a reference to the `MatChipInput` and putting the method on the directive instead. Then people would do something like this: `event.chipInput.clear()`

File: src/material/chips/chip-input.ts
Comment (mmalerba) [inline src/material/chips/chip-input.ts:145]: comment what this block is for

**Review by mmalerba — COMMENTED** [review]
Thanks for picking this back up, couple small things and it should be good to go.

File: src/material/chips/chip-input.ts
Comment (gioragutt) [reply src/material/chips/chip-input.ts:35]: Good idea! will change it 👍🏻

Comment (gioragutt) [reply src/material/chips/chip-input.ts:35]: What about the input itself? the `HTMLInputElement`... should it still be part of the event?

File: src/material/chips/chip-input.ts
Comment (gioragutt) [reply src/material/chips/chip-input.ts:148]: Added a comment, if you have a suggestion for how to make it shorter/clearer, I'll appreciate it 😅

File: src/material/chips/chip-input.ts
Comment (mmalerba) [reply src/material/chips/chip-input.ts]: We can't remove it without going through a deprecation period. You can mark it as deprecated:
```ts
/**
 * @deprecated use `chipInput.inputElement`
 * @breaking-change 13.0.0 remove this property.
 */
```

(You'll need to change `_inputElement` on `MatChipInput` to public and remove the underscore)

File: src/material/chips/chip-input.ts
Comment (mmalerba) [reply src/material/chips/chip-input.ts:148]: What you have sounds good to me, minor nit: "And" on the 3rd line doesn't need to be capitalized

**Review by mmalerba — COMMENTED** [review]
Does the MDC-based one need to be updated as well or does it share the same underlying code? I would think at least the tests for the MDC-based chips need to be updated to add the same tests that were added for the non-MDC chips

**Comment by gioragutt** [comment]
> Does the MDC-based one need to be updated as well or does it share the same underlying code? I would think at least the tests for the MDC-based chips need to be updated to add the same tests that were added for the non-MDC chips

@mmalerba time I worked on this you (I think) said that it was okay to not deal with the MDC implementation. If it is necessary now I'll take a look.

I'm pretty sure its not the same underlying code, but it's only an assumption. What say you? Should I take a look?

**Comment by mmalerba** [comment]
Ah, yeah since the time you started we've started to roll out the MDC-based components inside Google as an initial test before eventually moving them out of experimental. We'll need to make sure it works in the MDC one too so teams don't experience a regression as they migrate

**Comment by gioragutt** [comment]
Awesome. I’ll definitely take a look. Any helpful pointers for someone
who’s not seen MDC based components?

**Comment by mmalerba** [comment]
IIRC MDC didn't have built in support for connecting the chips with an input, so we just built that part on top, in Angular. That should mean that you don't have to worry too much about the MDC stuff going on in the background. 

If you do wind up needing to dive deeper into the MDC details, the ["advanced approach"](https://github.com/material-components/material-components-web/blob/master/docs/integrating-into-frameworks.md#the-advanced-approach-using-foundations-and-adapters) described in the MDC docs explains the approach we use to integration. And you can find the [code for the MDC chips here](https://github.com/material-components/material-components-web/tree/master/packages/mdc-chips).

**Comment by gioragutt** [comment]
> IIRC MDC didn't have built in support for connecting the chips with an input, so we just built that part on top, in Angular. That should mean that you don't have to worry too much about the MDC stuff going on in the background.
> 
> If you do wind up needing to dive deeper into the MDC details, the ["advanced approach"](https://github.com/material-components/material-components-web/blob/master/docs/integrating-into-frameworks.md?rgh-link-date=2021-02-16T20%3A56%3A54Z#the-advanced-approach-using-foundations-and-adapters) described in the MDC docs explains the approach we use to integration. And you can find the [code for the MDC chips here](https://github.com/material-components/material-components-web/tree/master/packages/mdc-chips?rgh-link-date=2021-02-16T20%3A56%3A54Z).

I'm looking at the code now, honestly it seems mostly copy-pasted, so it should be somewhat straightforward. I'm looking for any significant differences, so far it looks good.

File: src/material/chips/chip-input.ts
Comment (gioragutt) [reply src/material/chips/chip-input.ts:148]: Will change... I have this habit for a couple of years now when writing docs, I don't quite know why 🤦🏻‍♂️

**Comment by gioragutt** [comment]
Hey @mmalerba, finally got some time to work on it again.

I updated the mdc implementation, added tests to `mat-chip-grid` as requested, and took the liberty of cleaning up some of the code on the tests.

It's quite a bit of changes, so if it's too much, feel free to let me know, I'll revert whatever refactoring I did there.

**Review by mmalerba — APPROVED** [review]
LGTM. Thanks for all your work on this and for cleaning up the tests!

Just to explain a bit about our process, I'm going to mark this PR as ready to merge, but before we can actually merge it we'll need to run all of the tests for internal Google projects against this PR to make sure it doesn't break them. Feel free to ping the PR if you don't hear back about the results of that test run within a week.

**Comment by gioragutt** [comment]
> LGTM. Thanks for all your work on this and for cleaning up the tests!
> 
> Just to explain a bit about our process, I'm going to mark this PR as ready to merge, but before we can actually merge it we'll need to run all of the tests for internal Google projects against this PR to make sure it doesn't break them. Feel free to ping the PR if you don't hear back about the results of that test run within a week.

@mmalerba Amazing man, I'm glad to hear! TBH I just saw a comment of yours I missed, regarding the input on change event. Will it be okay if I fix it now?

**Comment by mmalerba** [comment]
I think it would actually be good to do the deprecation separately in a follow-up PR. Since this PR in its current state could go into a minor release, but we usually wait until major releases to deprecate things

**Comment by gioragutt** [comment]
> I think it would actually be good to do the deprecation separately in a follow-up PR. Since this PR in its current state could go into a minor release, but we usually wait until major releases to deprecate things

Gotcha. In that case - waiting for test results it is 😁

Thanks again for the time reviewing this, I really appreciate it and happy to contribute.

**Comment by gioragutt** [comment]
@mmalerba 🤓🏓

**Comment by mmalerba** [comment]
👍 I asked our caretaker to run the tests for this today

**Comment by gioragutt** [comment]
@mmalerba 🏓 any update?

**Comment by mmalerba** [comment]
It looks like there are a number of failures that will need to be investigated

**Comment by gioragutt** [comment]
Oh interesting! If there’s anything I can do to help investigate it, please
let me know!

**Comment by mmalerba** [comment]
So part of the problem appears to be people passing objects like `{input: ..., value: ...}` to functions that expect `MatChipInputEvent`. In order to make this change less breaking, I suggest changing `MatChipInputEvent` like this:

```ts
interface MatChipInputEvent {
  /** 
   * The native `<input>` element that the event is being fired for.
   * @deprecated Use `MatChipInputEvent#chipInput.inputElement` instead.
   * @breaking-change 13.0.0 This property will be removed.
   */
  input: HTMLInputElement;

  /** The value of the input. */
  value: string;
  
  /** 
   * Reference to the chip input that emitted the event. 
   * @breaking-change 13.0.0 This property will be made required.
   */
  chipInput?: MatChipInput;
}
```

In our examples, you can then just assert that `chipInput` is defined like this: `event.chipInput!.clear()`.

Lets try that and see how the tests go after

**Comment by gioragutt** [comment]
Got it, will try to find a couple of minutes to make the changes 👌🏻

**Comment by mmalerba** [comment]
I re-ran it with the fix applied and most of the errors went away. It looks like there's one more issue: When using the MDC chips the typing on `MatChipInputEvent#chipInput` is wrong because it refers to the non-MDC `MatChipInput`. The fix should be pretty straightforward. Instead of re-exporting the same interface from the non-MDC chips in the MDC package, just give MDC package its own version of the interface and export that

**Comment by gioragutt** [comment]
I’ll have to check that, it sounds too dumb 😂 I’ll be done with driving in
two hours or so, I’ll apply the fix then.

Thanks man!

**Comment by gioragutt** [comment]
> I re-ran it with the fix applied and most of the errors went away. It looks like there's one more issue: When using the MDC chips the typing on `MatChipInputEvent#chipInput` is wrong because it refers to the non-MDC `MatChipInput`. The fix should be pretty straightforward. Instead of re-exporting the same interface from the non-MDC chips in the MDC package, just give MDC package its own version of the interface and export that

Hey man, I looked at the code and I can't seem to find the behavior you're describing.

`src/material-experimental/mdc-chips/chip-input.ts` exports its own `MatChipInputEvent` interface.

**Comment by mmalerba** [comment]
Sorry about that, you're right. That project must've been doing something weird. I'll see if I can fix their app so we can get this in.

**Comment by gioragutt** [comment]
Hey @mmalerba 👋🏻 

And update on that?

**Comment by mmalerba** [comment]
My shift as the team's release manager starts tomorrow, I plan to look into it then

**Comment by mmalerba** [comment]
I ran the tests again after making some fixes to internal apps and everything passed, so I'll merge this today.

We currently have ~150 PRs that could be merged if they didn't cause failures of some kind in Google's codebase. For some of those cases the tests are finding legitimate issues with the PR, but for others it may just be a matter of the tests needing to be updated. That's why it can sometimes take a long time to merge PRs that look like they're ready to go on Github.

**Comment by mmalerba** [comment]
Thanks for the hard work on this! 🎉

**Comment by gioragutt** [comment]
> Thanks for the hard work on this! 🎉 

Thank you for helping me get this merged, your time is obviously invaluable, and I appreciate your responsiveness on this.

I've not experienced this in even smaller repos, so what you do is definitely not taken for granted.

Hope I'll find more opportunities to contribute!

**Comment by mmalerba** [comment]
@gioragutt Unfortunately I had to revert this. It broke a team that was using a custom version of the chips input. I'll ask them to investigate and fix it so we can roll forward

**Pull request merged**
